openedx/license-manager
# Read-only endpoints for subscriptions and licenses

## Description

* Exposes new routes for subscriptions and licenses (they're nested!):
```
/api/v1/subscriptions/?enterprise_customer_uuid={some_uuid} (subscription list endpoint)
/api/v1/subscriptions/{subscription_uuid}/ (subscription detail endpoint)
/api/v1/subscriptions/{subscription_uuid}/licenses/ (license list endpoint)
/api/v1/subscriptions/{subscription_uuid/licenses/{license_uuid}/ (license detail endpoint)
```

Link to the associated ticket: https://openedx.atlassian.net/browse/ENT-2750

## Testing considerations
Testing superuser flow (and most of the implementation):
- Create a subscription plan at http://localhost:18170/admin/subscriptions/subscriptionplan/ and allocate at least two licenses
- Navigate to http://localhost:18170/admin/subscriptions/license/ and change one license's status to 'active' or 'assigned', and give it an email: `fake@example.org`.
- Go to http://localhost:18170/api/v1/subscriptions/ and confirm the data looks correct. An important to field to notice is that `licenses` should have the correct number of total licenses, and one allocated license from the one changed in django admin
- Go to http://localhost:18170/api/v1/subscriptions/{subscription_uuid}/licenses/ and confirm the data is correct
  -  At this route, play around with the ordering and filtering parameters. For the most important one, add `?user_email=fake@example.org` and confirm that only the licenses with that email are returned
  - Add `?status=assigned` (assuming you chose `assigned` as the status for the license earlier) and confirm that only licenses with that status are returned
  - Add `?ordering=-user_email` and see the results ordered appropriately. The endpoint also supports ordering by status, `activation_date`, and `last_remind_date`. Those seemed like useful fields to be able to sort by in the future, but they are not explicitly called out in the ticket.
- Go to http://localhost:18170/api/v1/subscriptions/{subscription_uuid}/licenses/{license_uuid} and confirm that the data is correct for that license

Testing unauthenticated flow:
- Log out, hit all the endpoints listed above, and confirm they give a `401`

Testing non-staff flow:
- Log in as `verified@example.com` in the lms, hit all the endpoints listed above, and confirm they give a `403`

Testing staff flow:
- Make a user staff on license manager
- Hit `api/v1/subscriptions` with no query param and confirm no subscription plans show up
- Add a new subscription plan, set an `enterprise_customer_uuid`, and create a license to use with it
- Hit `api/v1/subscriptions/?enterprise_customer_uuid={enterprise_customer_uuid}` and confirm you can see only the subscription you specified. 
- HIt `api/v1/subscriptions/{subscription_uuid}/licenses/` and confirm you can see only licenses associated with that subscription

## Post-review

Squash commits into discrete sets of changes
